relay-server/src/actors/project_cache.rs

```diff
@@ -519,30 +615,90 @@ impl ProjectCacheBroker {
+    ) {
+        // The `Envelope` and `EnvelopeContext` will be dropped if the `Project::check_envelope()`
+        // function returns any error, which will also be ignored here.
+        let Some(Ok(checked)) = self
+            .projects
+            .get_mut(&project_key)
+            .map(|p| p.check_envelope(envelope, envelope_context)) else { return; };
+
+        if let Some((envelope, envelope_context)) = checked.envelope {
+            let mut process = ProcessEnvelope {
+                envelope,
+                envelope_context,
+                project_state: state.clone(),
+                sampling_project_state: None,
+            };
 
-        project.check_envelope(message.envelope, message.context)
+            if let Some(sampling_state) = sampling_state {
+                if state.organization_id == sampling_state.organization_id {
+                    process.sampling_project_state = Some(sampling_state)
+                }
+            }
+
+            EnvelopeProcessor::from_registry().send(process);
+        }
     }
 
+    /// Checks an incoming envelope and decides either process it immediately or buffer it.
+    ///
+    /// Few conditions are checked here:
+    /// - If there is no dynamic sampling key and the project is already cached, we do strait to
+    /// processing otherwise buffer the envelopes.
+    /// - If there is a dynamic sampling key is provided, if the root and sampling project
+    /// are cached - process the envelope, buffer otherwise.
+    ///
+    /// This means if the caches are hot we always process all the incoming envelopes without any
+    /// delay. But in case if the project state cannot be fetched, we keep buffering till the state
+    /// is everyntually updated.
+    ///
+    /// The flushing of the buffered envelopes happens in `update_state`.
     fn handle_validate_envelope(&mut self, message: ValidateEnvelope) {
-        // Preload the project cache for dynamic sampling in parallel to the main one.
-        if let Some(sampling_key) = utils::get_sampling_key(&message.envelope) {
-            self.get_or_create_project(sampling_key)
-                .prefetch(message.envelope.meta().no_cache());
-        }
+        let ValidateEnvelope { envelope, context } = message;
+
+        // Fetch the project state for our key and make sure it's not invalid.
+        let key = envelope.meta().public_key();
+        let project_state = self
+            .get_or_create_project(key)
+            .get_cached_state(envelope.meta().no_cache())
+            .filter(|st| !st.invalid());
+
+        // Also, fetch the project state for sampling key and make sure it's not invalid.
+        let sampling_key = utils::get_sampling_key(&envelope);
+        let sampling_state = sampling_key.and_then(|key| {
+            self.get_or_create_project(key)
+                .get_cached_state(envelope.meta().no_cache())
+                .filter(|st| !st.invalid())
+        });
```

Comment:
This can be moved inside `if let Some(state) = `.

Reply:
Bumping this comment. There's no need to `get_or_create_project` the sampling state when the `project_state` is `None`.

Reply:
In other places we use projects map directly, and from what I see this is the only one place now we call `self.get_or_create_project(...).get_cached_state(...)` which initiate the upstream request to update the sampling state if it's not in the cache. 

So if the `project_state` is `None` we are most probably request the update for it, and now just queue the envelope, and at the same time we request already sampling state to be able to process those queued envelopes when state are updated. 

But maybe I'm missing something.

Reply:
Sorry, you are right. I overlooked that `get_cached_state` triggers a fetch from the upstream.